src/streams/register_nack.rs

```diff
@@ -0,0 +1,560 @@
+
+/// Number of out of order packets we keep track of for reports
+/// A suggested maximum value min is MAX_DROPOUT / 2
+const MAX_MISORDER: u64 = 100;
+
+const U16_MAX: u64 = u16::MAX as u64 + 1_u64;
+
+/// The max number of NACKs we perform for a single packet
+const MAX_NACKS: u8 = 5;
+
+#[derive(Debug)]
+pub struct NackRegister {
+    /// Status of packets indexed by wrapping SeqNo.
+    packets: Vec<PacketStatus>,
+
+    /// Range of seq numbers considered NACK reporting.
+    active: Option<Range<SeqNo>>,
+}
+
+#[derive(Debug, Default, Clone, Copy)]
+struct PacketStatus {
+    received: bool,
+    nack_count: u8,
+}
+
+impl PacketStatus {
+    fn needs_nack(&self) -> bool {
+        !self.received && self.nack_count < MAX_NACKS
+    }
+
+    fn mark_received(&mut self) -> bool {
+        let new = !self.received;
+        self.received = true;
+        new
+    }
+
+    fn reset(&mut self) {
+        self.received = false;
+        self.nack_count = 0;
+    }
+}
+
+impl NackRegister {
+    pub fn new() -> Self {
+        NackRegister {
+            packets: vec![PacketStatus::default(); MAX_DROPOUT as usize],
+            active: None,
+        }
+    }
+
+    pub fn update(&mut self, seq: SeqNo) -> bool {
+        let Some(active) = self.active.clone() else {
+            // automatically pick up the first seq number
+            self.active = Some(seq..seq);
+            self.packet(seq).mark_received();
+            return true;
+        };
+
+        if seq < active.start {
+            // skip old seq numbers, report as not new
+            return false;
+        }
+
+        let mut next_active = active.start..active.end.max(seq);
+
+        let new = self.packet(seq).mark_received();
+
+        next_active.start = {
+            let min = next_active.end.saturating_sub(MAX_MISORDER);
+            let mut start = (*next_active.start).max(min);
+            while start < *next_active.end {
+                if !self.packet(start.into()).received {
+                    break;
+                }
+                start += 1;
+            }
+            start.into()
+        };
+
+        // reset packets that are rolling our of the nack window
+        for s in *active.start..*next_active.start {
+            let p = self.packet(s.into());
+            if !p.received {
+                debug!("Seq no {} missing after {} attempts", seq, p.nack_count);
+            }
+            self.packet(s.into()).reset();
+        }
+
+        self.active = Some(next_active);
+
+        new
+    }
+
+    pub fn max_seq(&self) -> Option<SeqNo> {
+        Some(self.active.as_ref()?.end)
+    }
+
+    /// Create a new nack report
+    ///
+    /// This modifies the state as it counts how many times packets have been nacked
+    pub fn nack_report(&mut self) -> Option<Vec<Nack>> {
+        let active = self.active.as_ref()?;
+
+        if active.is_empty() {
+            return None;
+        }
+
+        let mut nacks = vec![];
```

Comment:
`ReportList` is designed to have a fixed size so it seems wasteful to allocate here. Instead we can use `ReportList::new` and then `ReportList::push`.

Also the result of this method gets pushed onto a `&mut VecDeque<Rtcp>` that is allocated a single time in `Session`. We could entirely avoid allocating here we changed the signature to

```rust
pub fn nack_report(&mut self, on_nack: F)
where 
  F: FnMut(nack: Nack)
```

Reply:
Since this is tightly coupled to the RTCP type, I don't mind making it produce the specific type directly.

Reply:
So

```rust
pub fn nack_report(&mut self, on_nack: F)
where 
  F: FnMut(rtcp: Rtcp)
```
?

Reply:
Either the way we collect feedback through the rest of handle_timeout

```rust
pub fn nack_report(&mut self, feedback: &mut VecDeque<Rtcp>) { … }
```

or

```rust
pub fn nack_report(&mut self) -> Option<Rtcp> { … }
```

Reply:
Aight, I guess the former is slightly more flexible since `nack_report` could, in theory, produce two instances of `Rtcp`. But I guess we might not want to cause that much NACKing anyway